### PR TITLE
Adding markers to workbench plots

### LINKED ISSUE #26657: Adding vertical and horizontal markers to workbench plots
Plots in the workbench should have an option to add vertical/horizontal dragable lines.

To place: `Right click->Add marker->Horizontal/Vertical`
To move: `Left click` and drag

### PR DESCRIPTION
**Pull request by BassoLorenzo**
**Description of work.**
This PR adds the following to the plots in Workbench:
 - vertical and horizontal markers
 - labels for the markers
 - editing menu for label, line style and position
 - exporting markers to table workspaces
 - importing markers from table workspaces

**To test:**
Code review.

1. Open a plot in workbench
2. `right click->Markers->Horizontal/Vertical`, a green dashed line should appear with label `marker 0`
3. Add more markers, the names should increase progressively (i.e. `marker 0`, `marker 1` etc.)
4. Hover the cursor over one, the cursor should become an open hand
5. Click and drag, the cursor should become a closed hand. The label should disappear and the marker should move.
6. Release the marker, the label should appear again in the correct position
7. Pan/Zoom, the label should always be in the correct position
8. Hover the cursor over a marker and right click.
9. `marker 0->Edit`, an edit window will pop up
10. Adjust the position and change the name, the marker should move and the label should be updated
11. Right click again
12. `<name>->Delete`, marker and label should disappear
13. Now add a vertical and horizontal marker
14. Hover above the intersections and drag them. They should both move
15. Right click above the intersection. Now there should be 2 menu, one for each marker.
16. Double click a marker, the editing menu will appear.
17. The labels should not trail off the edge of the figure or lay across the marker, try using long names and labels like `$exp(-\frac{(x-\mu)^2}{2 \sigma^2})\quad \aleph$`

Fixes #26657.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by Harrietbrown** [comment]
A couple of things I've found

1. On horizontal lines if the marker label is too long it trails off the side of the plot, the vertical marker doesn't have this issue, you should make it so that the marker text on the horizontal line is aligned right so it can't trail off.
1. The range that you can drag the vertical line is different to the range that the mouse changes. So you can hover over the vertical line with the open hand mouse button but not be able to drag and drop
1. Double clicking to open the editing window where two lines cross over opens both in sequence, not sure if that is the best way to handle it.
1. On the open hand cursor the pointer bit is in the centre, so it's not always clear where to position the mouse for the crossover point (I hope that makes sense).
1. If you change marker label into something with two separate words ("foo bar") it raises an error when it expects the second word to be convertible to an int.

**Comment by Mantid-Matthew** [comment]
I've had a look at this and in general is working really well. I have a few minor comments though.
  * When I tried this in python3 I ran into a few errors, this seems to be the standard thing of dict.keys() returning an iterator not a list like it does in python2
  * We are trying to not offer options in menus e.c.t. unless they are strictly needed, is there a specific use case where the import export options are required to be on the menu?
  * There seems to be an odd interaction with the save to png option whereby all the markers even ones I had previously deleted were saved producing a plot like the below
test2

File: qt/applications/workbench/workbench/plotting/figureinteraction.py [hunk 1]
```diff
@@ -55,13 +58,22 @@ def __init__(self, fig_manager):
...
 
         self.canvas = canvas
         self.toolbar_manager = ToolbarStateManager(self.canvas.toolbar)
+        self.toolbar_manager.home_button_connect(self.redraw_annotations)
         self.fit_browser = fig_manager.fit_browser
         self.errors_manager = FigureErrorsManager(self.canvas)
+        self.markers = []
+        self.vertical_markers = []
+        self.valid_lines = [str(name) for name in ['solid', 'dashed', 'dotted', 'dashdot']]
+        self.valid_colors = {'blue': 'C0', 'orange': 'C1', 'green': 'C2', 'red': 'C3', 'purple': 'C4'}
```
Comment (Mantid-Matthew) [inline qt/applications/workbench/workbench/plotting/figureinteraction.py]: Just as a code style thing could we define these as constants at the top of the file.

File: qt/applications/workbench/workbench/plotting/figureinteraction.py [hunk 2]
```diff
@@ -110,13 +163,45 @@ def move_and_show(editor):
...
+        QApplication.restoreOverrideCursor()
+        marker_menu = QMenu(marker.name, menu)
+        marker_action_group = QActionGroup(marker_menu)
+
+        delete = marker_menu.addAction("Delete", lambda: self._delete_marker(marker))
+        edit = marker_menu.addAction("Edit", lambda: self._edit_marker(marker))
+
+        for action in [delete, edit]:
+            marker_action_group.addAction(action)
+            action.setCheckable(True)
```
Comment (Mantid-Matthew) [inline qt/applications/workbench/workbench/plotting/figureinteraction.py]: Out of curiosity why do we need to set the action as checkable here?

File: qt/applications/workbench/workbench/plotting/test/test_figureinteraction.py [hunk 3]
```diff
@@ -131,12 +140,15 @@ def test_right_click_gives_context_menu_for_plot_without_fit_enabled(self, mocke
...
                     self.assertEqual(expected_qmenu_calls, mocked_qmenu_cls.call_args_list)
                     # 4 actions in Axes submenu
                     self.assertEqual(4, qmenu_call2.addAction.call_count)
                     # 2 actions in Normalization submenu
                     self.assertEqual(2, qmenu_call3.addAction.call_count)
+                    # 2 actions in Markers submenu
+                    self.assertEqual(2, qmenu_call3.addAction.call_count)
 
     def test_toggle_normalization_no_errorbars(self):
         self._test_toggle_normalization(errorbars_on=False, plot_kwargs={'distribution': True})
```
Comment (Mantid-Matthew) [inline qt/applications/workbench/workbench/plotting/test/test_figureinteraction.py]: Could we add some tests to do with adding and removing markers from the plot?

**Review by Matthew-Andrew — CHANGES_REQUESTED** [review]
In general looks really good and the new functionality works really nicely :+1: I've added a few comments above but in general great job.

**Review by Matthew-Andrew — DISMISSED** [review]
I'm now happy with this. Seems to work really nicely good job :+1:

**Review by martyngigg — CHANGES_REQUESTED** [review]
This is a really great addition.

I only have one comment that I would prefer we address now. We have been trying to get better at ensuring that the method/function documentation exists and is up to date. Can we make sure that all of the new methods have documentation and that any methods/functions that have been changed have their docs updated?

**Comment by martyngigg** [comment]
It also looks like we have a conflict now. Can we rebase with master?

**Comment by martyngigg** [comment]
This looks good. Thanks. Just a documentation warning left.

File: docs/source/workbench/plotwindow.rst [hunk 4]
```diff
@@ -27,10 +27,20 @@ Plot Window Context Menu
 ------------------------
 
 .. image:: ../images/Workbench/PlotWindow/PlotWindowContextMenu.png
-    :scale: 70%
+    :height: 400px
+
+Right-clicking within your plot will open the plot context menu.
+From this menu you can:
+- set the axes' scales to linear or logarithmic.
```
Comment (martyngigg) [inline docs/source/workbench/plotwindow.rst]: The formatting here doesn't look quite right here as this is not formatted as a list. I think you just need a blank line between the line first line with the `-` and the previous line.

File: docs/source/workbench/plotwindow.rst [hunk 5]
```diff
@@ -31,12 +31,13 @@ Plot Window Context Menu
 
 Right-clicking within your plot will open the plot context menu.
 From this menu you can:
-- set the axes' scales to linear or logarithmic.
```
Comment (martyngigg) [inline docs/source/workbench/plotwindow.rst]: Sorry to be picky. We need a single blank line above this line or the list is formatted as pre-formatted block (note the line down the left-hand side):
Screenshot from 2019-08-23 16-22-53

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
OSX failure is unrelated. Documentation warnings are fixed in #26712 after move to Ubuntu 18.04 on PR builds.

**Pull request merged**